Outbound/inbound message handling; compilation fixes

dr7ana opened the pull request.

dr7ana committed:
initial commit for next PR

tewinget committed:
implement sending "path control" (onioned control) message

TODO: handle at transit hop, handle at terminal hop, create and handle responses

dr7ana committed:
introset and message transmission underway

- message handling through classes that inherit from PathSet
- cleanups around link_manager
- etc etc

dr7ana committed:
Outbound context absorbed sendcontex

- message transmission routed through refactored handling
- still work to be done, but now to make it compile at least

jagerman committed:
Bring oxen-encoding back in as a submodule

As long as it's loaded before oxen-mq or libquic, this should end up
being the one that gets used in oxen-mq/libquic, and so we don't have to
update those when we want something in a newer oxen-encoding.

dr7ana committed:
Compilation fixes

- almost all errors have been commented out for refactor or already refactored
- committing this prior to sorting out the cmake structure
- upcoming include-what-you-use application

dr7ana committed:
Review commit

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -13,16 +12,38 @@
+#include "CLI/Option.hpp"
+#include "CLI/impl/App_inl.hpp"
+#include "CLI/impl/Option_inl.hpp"
+#include "fmt/core.h"
```
The includes above here (i.e. headers that aren't part of llarp) should be <...> instead of "...".

jagerman replied:
Actually I think *most* of them should be.

The rule we decided on for this a couple years ago was to use `<a/b>` for a fully qualified headers -- certainly for external library headers, but also for fully-qualified llarp headers (e.g. `#include <llarp/util/logging.hpp>` rather than `#include "llarp/util/logging.hpp"`).

For includes in the same directory or a subdirectory, though, we use double-quotes, which try relative paths first (which angle includes do not).  So for instance, inside `llarp/foo.cpp` we would prefer `"#include "bar.hpp"` to `#include <llarp/bar.hpp>`, and likewise for includes from subdirectories, i.e. `#include "util/logging.hpp"` instead of `#include <llarp/util/logging.hpp>`.

jagerman replied:
(Personally I'm fine with `#include "../whatever.hpp"` as well, e.g. so `llarp/foo/bar.hpp` could `#include "../util/logging.hpp"` rather than `#include <llarp/util/logging.hpp>` but Jeff and Tom vetoed that idea)

dr7ana replied:
Yep you got it. These were all changed by IWYU haha, I was probably just going to write a little script to fix then manually verify rather than changing everything by hand

jagerman committed:
"Refactor" aka delete Crypto/CryptoManager

- Get rid of CryptoManager.
- Get rid of Crypto.
- Move all the Crypto instance methods to llarp::crypto functions.
  (None of them needed to be methods at all, so this is simple).
- Move sodium/ntru initialization into static initialization.
- Add llarp::csrng, which is an available llarp::CSRNG instance which is
  a bit easier than needing to construct a `CSRNG rng{};` in various
  places.
- Various related small simplifications/cleanups.

jagerman commented on llarp/CMakeLists.txt:
```diff
@@ -160,280 +243,140 @@ add_library(lokinet-layer-onion
-# helper function to link a library to lokinet-base, enable lto, add to lokinet-amalgum and then link to other libs
-function(lokinet_link_lib libname)
-  message(DEBUG "created target: ${libname}")
-  enable_lto(${libname})
```
This commit is currently dropping the `enable_lto` that gets applied to each library, but that needs to be brought back (LTO significantly reduces the size of release binaries, and optimizes performance as well).

We could do it in a different way, though.  Perhaps something something:

```cmake
# Add an internal lokinet static library target, enables LTO (if enabled) on the target,
# and links it to the common lokinet-base interface.
# Invoke with the target/library name (e.g. "lokinet-foo") and list of source files, e.g.
#     lokinet_add_library(lokinet-foo foo/source1.cpp foo/source2.cpp)
function(lokinet_add_library name)
    add_library(${libname} STATIC ${ARGN})
    target_link_libraries(${libname} PUBLIC lokinet-base)
    enable_lto(${libname})
endfunction()
```
and then changing all the `add_library(lokinet-foo STATIC sources...)` to `lokinet_add_library(lokinet-foo sources...)` to DRY out these two common additions.

dr7ana replied:
Cool, I will loop in `target_link_libraries(${libname} INTERFACE lokinet-amalgum)` as well

tewinget commented on daemon/lokinet.cpp:
```diff
@@ -350,8 +371,8 @@ namespace
   {
-    if (auto result = Lokinet_INIT())
-      return result;
+    // if (auto result = Lokinet_INIT())
```
Just remove; appears all this function did is `return 0`...

tewinget commented on llarp/CMakeLists_old.txt:
```diff
@@ -0,0 +1,421 @@
+include(Version)
```
This file probably shouldn't be committed.

tewinget commented on llarp/config/key_manager.cpp:
```diff
@@ -93,23 +93,23 @@ namespace llarp
       auto identityKeygen = [](llarp::SecretKey& key) {
         // TODO: handle generating from service node seed
-        llarp::CryptoManager::instance()->identity_keygen(key);
+        llarp::crypto::identity_keygen(key);
```
drop `llarp::`?

tewinget commented on llarp/config/key_manager.cpp:
```diff
@@ -93,23 +93,23 @@ namespace llarp
     // load encryption key
     auto encryptionKeygen = [](llarp::SecretKey& key) {
-      llarp::CryptoManager::instance()->encryption_keygen(key);
+      llarp::crypto::encryption_keygen(key);
```
drop `llarp::`?

tewinget commented on llarp/crypto/crypto.cpp:
```diff
@@ -113,28 +108,8 @@ namespace llarp
-
   std::optional<AlignedBuffer<32>>
-  Crypto::maybe_decrypt_name(std::string_view ciphertext, SymmNonce nounce, std::string_view name)
+  crypto::maybe_decrypt_name(std::string_view ciphertext, SymmNonce nounce, std::string_view name)
```
`s/nounce/nonce/`

tewinget commented on llarp/crypto/crypto.hpp:
```diff
@@ -17,12 +15,8 @@ namespace llarp
-
     /// decrypt cipherText given the key generated from name
     std::optional<AlignedBuffer<32>>
     maybe_decrypt_name(std::string_view ciphertext, SymmNonce nounce, std::string_view name);
```
`s/nounce/nonce/`

tewinget commented on llarp/crypto/types.cpp:
```diff
@@ -2,14 +2,10 @@
-#include <sodium/crypto_sign.h>
-#include <sodium/crypto_sign_ed25519.h>
-#include <sodium/crypto_scalarmult_ed25519.h>
+// #include <sodium/crypto_hash_sha512.h>
```
remove if not needed?

tewinget commented on llarp/dns/dns.hpp:
```diff
@@ -1,30 +1,29 @@
 #pragma once
 
+#include <llarp/util/logging.hpp>
```
I think IWYU got this one wrong...can't see how this header would need included here.

dr7ana committed:
Cmake restructuring

dr7ana committed:
Clang format include sorting + CMake

- includes are now sorted in consistent, logical order; first step in an attempt to fix the tomfoolery (no relation to Tom) brought in by include-what-you-use
- shuffled around some cmake linking to simplify dependency graph
- superfluous files removed

jagerman commented on .clang-format:
```diff
@@ -55,7 +55,20 @@ PointerAlignment: Left
+#   - System dependencies
+IncludeBlocks: Regroup
+IncludeCategories:
+  - Regex: '((")(.)+(.hpp"))'
+    Priority: 2
+  - Regex: '((<)(llarp)(/))'
+    Priority: 3
+  - Regex: '((<)[^(llarp)](.)+(.h))'
+    Priority: 4
+  - Regex: '((<)(\w)+(>))'
+    Priority: 5
```
```suggestion
  - Regex: '^".+\.h(pp)?"$'  # "quoted" includes of local headers
    Priority: 2
  - Regex: '^<llarp[./]'     # fully qualified llarp/ includes (and <llarp.hpp>)
    Priority: 3
  - Regex: '^<.*\.h(pp)?>$'  # external library includes
    Priority: 4
  - Regex: '^<\w+>$'         # stl includes
    Priority: 5
```

jagerman replied:
Updated the second one so that `<llarp/...>` and `<llarp.hpp>` get grouped.

jagerman commented on cmake/CMakeGraphVizOptions.cmake:
```diff
@@ -0,0 +1,6 @@
+set(GRAPHVIZ_GRAPH_NAME "graph.dot" CACHE STRING "")
+# set(GRAPHVIZ_MODULE_LIBS OFF CACHE BOOL "")
+# set(GRAPHVIZ_EXTERNAL_LIBS OFF CACHE BOOL "")
```
Should these two be removed?

jagerman commented on crypto/libntrup/src/ntru.cpp:
```diff
@@ -1,5 +1,7 @@
 #include <libntrup/ntru.h>
 
+#include "libntrup/ntru_api.h"
```
Should be angle brackets

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet-bootstrap.cpp:
```diff
@@ -1,26 +1,17 @@
 #include <openssl/x509.h>
 #endif
 
-#include <sstream>
```
Dropping this seems suspicious, because we're using a stringstream in this file

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet-vpn.cpp:
```diff
@@ -1,25 +1,19 @@
 
 #ifdef _WIN32
 // add the unholy windows headers for iphlpapi
-#include <winsock2.h>
```
`winsock2.h` is a special holy water include that changes other includes based on when it gets included, and so generally *has* to be before other Windows-specific includes.  Yay Windows.

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
-#include <llarp/util/str.hpp>
+#include <llarp/util/lokinet_init.h>
+
+#include <bits/chrono.h>
```
Shouldn't be here; this is an stl-internal header; we should have `<chrono>` instead.

dr7ana replied:
fucking iwyu

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
 #include <llarp/util/service_manager.hpp>
 #endif
 
+#include "CLI/Error.hpp"
+#include "CLI/impl/App_inl.hpp"
+#include "CLI/impl/Option_inl.hpp"
+#include "CLI/Option.hpp"
+#include "llarp/constants/files.hpp"
+#include "llarp/constants/platform.hpp"
+#include "llarp/ev/ev.hpp"
+#include "llarp/util/logging.hpp"
+#include "llarp/util/thread/threading.hpp"
+#include "oxen/log.hpp"
+#include "oxen/log/catlogger.hpp"
+#include "oxen/log/level.hpp"
+#include "oxen/log/ring_buffer_sink.hpp"
+#include "oxen/log/type.hpp"
```
Angle includes

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
 #endif
 
+#include "CLI/Error.hpp"
+#include "CLI/impl/App_inl.hpp"
+#include "CLI/impl/Option_inl.hpp"
```
Internal headers that shouldn't be here

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
+#include "oxen/log.hpp"
+#include "oxen/log/catlogger.hpp"
+#include "oxen/log/level.hpp"
+#include "oxen/log/ring_buffer_sink.hpp"
```
Shouldn't be here

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
+#include "llarp/util/logging.hpp"
+#include "llarp/util/thread/threading.hpp"
+#include "oxen/log.hpp"
+#include "oxen/log/catlogger.hpp"
+#include "oxen/log/level.hpp"
```
These are implicit from `<oxen/log.hpp>`

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
+#include "oxen/log/catlogger.hpp"
+#include "oxen/log/level.hpp"
+#include "oxen/log/ring_buffer_sink.hpp"
+#include "oxen/log/type.hpp"
```
`oxen/log/type.hpp` is implict from `oxen/log.hpp`

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
+
+#include <CLI/CLI.hpp>
+
+#include "fmt/core.h"
```
<>

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
+
+#include <CLI/CLI.hpp>
+
+#include "fmt/core.h"
+#include "spdlog/common.h"
```
Shouldn't be here

dr7ana replied:
IWYU

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
+#include <exception>
+#include <filesystem>
+#include <future>
+#include <initializer_list>
```
`-`

dr7ana replied:
IWYU

jagerman commented on llarp/util/easter_eggs.cpp:
```diff
@@ -1,39 +0,0 @@
-
-// i heckin love static initalization
-WineKiller lmao{};
-#endif
```
The above should come back, but should move into `/daemon/lokinet.cpp` (so that it only fires on full lokinet, but not embedded lokinet).

jagerman replied:
(It doesn't need to be in a struct; it could easily just be in the `main()` function)

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -1,28 +1,52 @@
+#include "CLI/impl/App_inl.hpp"
+#include "CLI/impl/Option_inl.hpp"
+#include "CLI/Option.hpp"
+#include "llarp/constants/files.hpp"
+#include "llarp/constants/platform.hpp"
+#include "llarp/ev/ev.hpp"
+#include "llarp/util/logging.hpp"
+#include "llarp/util/thread/threading.hpp"
+#include "oxen/log.hpp"
```
angle brackets

dr7ana replied:
IWYU

jagerman committed:
Add format.sh detection of include problems

jagerman commented on llarp/crypto/types.hpp:
```diff
@@ -1,10 +1,13 @@
 #include <llarp/util/fs.hpp>
+#include <llarp/util/types.hpp>
+
+#include <sodium.h>
```
This is putting sodium headers in nearly every compilation unit across the codebase, whereas sodium is meant to be just an internal implementation detail of the .cpp.

jagerman commented on llarp/crypto/crypto.cpp:
```diff
@@ -1,26 +1,13 @@
-#include <sodium/crypto_core_ed25519.h>
-#include <sodium/crypto_aead_xchacha20poly1305.h>
-#include <sodium/randombytes.h>
-#include <sodium/utils.h>
```
Breaking out the sodium headers this way was deliberate to only include the parts of libsodium we are using, rather than all of libsodium.

jagerman commented on llarp/endpoint_base.hpp:
```diff
@@ -1,29 +1,27 @@
-  namespace quic
-  {
-    class TunnelManager;
-  }
```
This should remain as a forward declaration (include of having the `#include <llarp/link/tunnel.hpp>`), so that lokinet-core-utils doesn't have to depend on lokinet-wire (where `link/tunnel.cpp` gets compiled).

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -651,6 +632,17 @@ main(int argc, char* argv[])
+      static const char* text = "Don't run lokinet in wine, aborting startup";
+      static const char* title = "Lokinet Wine Error";
+      MessageBoxA(NULL, text, title, MB_ICONHAND);
+      abort();
```
```suggestion
      std::abort();
```
(This is a bug from previous code this was copied from)

jagerman commented on llarp/nodedb.cpp:
```diff
@@ -243,7 +238,7 @@ namespace llarp
   NodeDB::num_loaded() const
   {
-    return router.loop()->call_get([this]() { return entries.size(); });
+    return router.loop()->call_get([this]() -> size_t { return entries.size(); });
```
This change doesn't seem necessary/wanted

jagerman commented on llarp/pow.cpp:
```diff
@@ -1,7 +1,6 @@
 #include "pow.hpp"
```
I think we should just nuke this as I don't think it does anything useful.

dr7ana committed:
Review fixes

- cleaned up include-what-you-use errors (mostly quoted includes)
- misc fixes

jagerman commented on llarp/service/outbound_context.hpp:
```diff
@@ -10,29 +11,82 @@
+
     void
-    Tick(llarp_time_t now) override;
+    Tick(std::chrono::milliseconds now) override;
```
This (any many other in this file) should remain as `llarp_time_t` as they are representing timepoints (and even though `llarp_time_t` is a typedef to `ms`, it's useful also signals that a value is a time point, rather than a duration.  `llarp_time_t` isn't well implemented in that it really should be a proper time point, but that's a refactor for another day, and switching it to a plain duration is a step in the wrong direction here.

Looking at some of the others, it's possible that one or two of them should be converted when they are conceptually durations (and not conceptually time points), such as RTT().

jagerman commented on llarp/service/protocol.cpp:
```diff
@@ -39,26 +39,22 @@ namespace llarp::service
-    if (!BEncodeMaybeReadDictInt("a", proto, read, k, buf))
-      return false;
+    // if (!BEncodeMaybeReadDictInt("a", proto, read, k, buf))
+    //   return false;
```
Is this to be replaced or to be removed?

dr7ana replied:
unsure, i thought we discussed this proto as a member variable being replaced or re-implemented in some way? also aren't these objects going the way of the dinosaurs like the other message objects

maybe i'm falsely remembering, will mention this tomorrow in standup

jagerman commented on llarp/service/protocol.cpp:
```diff
@@ -69,7 +65,7 @@ namespace llarp::service
     try
     {
-      btdp.append("a", static_cast<uint64_t>(proto));
+      // btdp.append("a", static_cast<uint64_t>(proto));
```
replace or remove?

dr7ana replied:
^^ above

jagerman commented on llarp/service/protocol.cpp:
```diff
@@ -105,19 +98,18 @@ namespace llarp::service
     try
     {
-      btdp.append("a", static_cast<uint64_t>(proto));
+      // btdp.append("a", static_cast<uint64_t>(proto));
```
Replace or remove?

dr7ana replied:
^^ same as above

jagerman commented on llarp/vpn/android.hpp:
```diff
@@ -1,12 +1,13 @@
+
 #include <llarp.hpp>
 
+#include <stdio.h>
```
```suggestion
#include <cstdio>
```

jagerman commented on pybind/llarp/router.cpp:
```diff
@@ -1,8 +1,8 @@
 #include <llarp/router/router.hpp>
 #include <llarp/tooling/hive_router.hpp>
 
+#include <common.hpp>
```
```suggestion
#include "common.hpp"
```

dr7ana replied:
this change errors

jagerman commented on pybind/llarp/router_id.cpp:
```diff
@@ -1,7 +1,7 @@
-
 #include <llarp/router_id.hpp>
 
+#include <common.hpp>
```
```suggestion
#include "common.hpp"
```

dr7ana replied:
this change errors

jagerman commented on pybind/llarp/service/address.cpp:
```diff
@@ -1,6 +1,7 @@
-#include <common.hpp>
 #include <llarp/service/address.hpp>
 
+#include <common.hpp>
```
```suggestion
#include "common.hpp"
```

dr7ana replied:
this change errors

jagerman commented on pybind/llarp/tooling/router_event.cpp:
```diff
@@ -1,14 +1,13 @@
 
-#include <llarp/messages/relay_status.hpp>
-#include <llarp/path/path.hpp>
+#include <common.hpp>
```
```suggestion
#include "common.hpp"
```

dr7ana replied:
this change errors

jagerman commented on pybind/llarp/tooling/router_hive.cpp:
```diff
@@ -1,10 +1,10 @@
+#include <llarp/router/router.hpp>
+#include <llarp/tooling/router_hive.hpp>
+
 #include <common.hpp>
```
```suggestion
#include "common.hpp"
```

dr7ana replied:
this change errors

jagerman reviewed (commented):
Done file-by-file review.

jagerman committed:
Fix systemd linkage

jagerman commented:
I pushed some changes here to bump the oxen-encoding submodule (to fix `require` not working depending on what's installed on the system), and a fix for the systemd linking.

Test code is still broken; I think we should probably just disable that from CI as a temporary measure.

jagerman committed:
Temporarily disable building/running test suite

dr7ana committed:
CI Fixes, squash

dr7ana committed:
windows fixes

dr7ana committed:
squash

jagerman committed:
Disable Router Hive & DEB builds

They aren't useful right now.

jagerman committed:
Default libcrypt to whether or not we find it

Also deliberately don't provide the function at all if we aren't
compiled with libcrypt so that we can't link if we try to call it when
not available.

jagerman committed:
Bump WinDivert to latest

There are some bugfix releases, but also the upstream hash was wrong for
the current release anyway (upstream changed it?).

dr7ana committed:
Bump libzmq static; remove patches

dr7ana committed:
Windows CI fixes

- some weird function call business
- string formatting
- etc

jagerman committed:
Bump oxen-mq for libzmq llvm static build fix

jagerman committed:
Don't export CC

This was causing cmake to double-invoke ccache for compilation units,
and broke oxen-mq's local zmq build for the same reason.

dr7ana committed:
Weird clang unused variable CI fix

jagerman committed:
Bump armhf static build to bullseye

Use backports for cmake because ngtcp2 needs 3.20+.

jagerman committed:
Turn off terrible new cmake 3.24+ default

jagerman committed:
Remove unused file

jagerman committed:
Give up for now

jagerman committed:
Bump libquic to latest for static build fixes
